packages/core/expressions/src/monaco.ts

```diff
@@ -0,0 +1,174 @@
+    rules: [
+      { token: 'keyword', foreground: '#003694', fontStyle: 'bold' },
+      { token: 'operators', foreground: '#003694', fontStyle: 'bold' },
+      { token: 'string', foreground: '#009966' },
+      { token: 'number', foreground: '#009966' },
+      { token: 'variable', foreground: '#006699' },
+      { token: 'brackets', foreground: '#993399' },
+    ],
+    colors: {
+      'editor.foreground': '#000000',
+    },
+  })
+}
+
+export const registerLanguage = (schema: NamedSchemaDefinition) => {
+  const languageId = getLanguageId(schema.name)
+
+  if (monaco.languages.getEncodedLanguageId(languageId) !== 0) {
+    console.warn(`Language "${languageId}" has already been registered`)
+    return { languageId }
+  }
+
+  const tokenTree = buildTokenTree(schema.definition)
+
+  console.debug(`Registering language "${languageId}" with token tree:`, tokenTree)
+
+  const keywords = ['not', 'in', 'contains']
+
+  monaco.languages.register({ id: languageId })
+
+  monaco.languages.setMonarchTokensProvider(languageId, {
+    keywords,
+    tokenizer: {
+      root: [
+        [/[a-zA-Z][\w$]*/, {
+          cases: {
+            '@keywords': 'keyword',
+            '@default': 'variable',
+          },
+        }],
+        [/==|!=|~|\^=|=\^|>=?|<=?|&&|\|\|/, 'operators'],
+        [/[()]/, 'brackets'],
+        [/".*?"/, 'string'],
+        [/\d+/, 'number'],
+      ],
+    },
+  } as MonarchLanguage)
+
+  monaco.languages.registerCompletionItemProvider(languageId, {
+    provideCompletionItems: (model, position) => {
+      const word = model.getWordUntilPosition(position)
+      const range = {
+        startLineNumber: position.lineNumber,
+        endLineNumber: position.lineNumber,
+        startColumn: word.startColumn,
+        endColumn: word.endColumn,
+      }
+
+      const suggestions = keywords.map((key) => ({
+        label: key,
+        kind: monaco.languages.CompletionItemKind.Keyword,
+        insertText: key,
+        range,
+      }))
+
+      return { suggestions }
+    },
+  })
+
+  monaco.languages.registerCompletionItemProvider(languageId, {
+    // run the following function when a period is typed
+    triggerCharacters: ['.'],
+
+    // function to generate object autocompletion
+    provideCompletionItems: (model, position) => {
+      const word = model.getWordUntilPosition(position)
+
+      const range = {
+        startLineNumber: position.lineNumber,
+        endLineNumber: position.lineNumber,
+        startColumn: word.startColumn,
+        endColumn: word.endColumn,
+      }
+
+      // split everything the user has typed on the current line up at each space, and only look at the last word
+      const lastChars = model.getValueInRange({
+        startLineNumber: position.lineNumber,
+        startColumn: 0,
+        endLineNumber: position.lineNumber,
+        endColumn: position.column,
+      })
+
+      // nosemgrep: javascript.lang.security.audit.incomplete-sanitization.incomplete-sanitization
```

Comment:
nit: we can remove this
```suggestion
```

Reply:
Thanks for reminding me of this. I've removed this line